File: application/backend/inventory/services/order_status.py
from django.db import transaction
from rest_framework.exceptions import ValidationError
from inventory.models import Notification, Order, OrderStatusHistory
from inventory.services.notifications import create_notification

ALLOWED_STATUS_TRANSITIONS = {
    Order.Status.PENDING: {
        Order.Status.PROCESSING,
        Order.Status.CANCELLED,
    },
    Order.Status.PROCESSING: {
        Order.Status.SHIPPED,
        Order.Status.CANCELLED,
    },
    Order.Status.SHIPPED: {Order.Status.DELIVERED},
    Order.Status.DELIVERED: set(),
    Order.Status.CANCELLED: set(),
}

@transaction.atomic
def transition_order_status(order, new_status, *, performed_by=None):
    locked_order = Order.objects.select_for_update().get(pk=order.pk)
    allowed_statuses = ALLOWED_STATUS_TRANSITIONS[locked_order.status]
    if new_status not in allowed_statuses:
        raise ValidationError(
            {
                "status": (
                    f"Cannot transition order from '{locked_order.status}' "
                    f"to '{new_status}'."
                )
            }
        )
    previous_status = locked_order.status
    locked_order.status = new_status
    locked_order.save(update_fields=("status", "updated_at"))
    OrderStatusHistory.objects.create(
        order=locked_order,
        from_status=previous_status,
        to_status=new_status,
        performed_by=performed_by,
    )
    notification_event = {
        Order.Status.SHIPPED: Notification.EventType.ORDER_SHIPPED,
        Order.Status.DELIVERED: Notification.EventType.ORDER_DELIVERED,
    }.get(new_status)
    if notification_event:
        create_notification(
            user=locked_order.user,
            order=locked_order,
            event_type=notification_event,
        )
    return locked_order

File: application/frontend/src/lib/order-status.ts
import { OrderStatus } from "@/types";

const transitions: Record<OrderStatus, OrderStatus[]> = {
  [OrderStatus.Pending]: [OrderStatus.Processing, OrderStatus.Cancelled],
  [OrderStatus.Processing]: [OrderStatus.Shipped, OrderStatus.Cancelled],
  [OrderStatus.Shipped]: [OrderStatus.Delivered],
  [OrderStatus.Delivered]: [],
  [OrderStatus.Cancelled]: [],
};

export function allowedStatusTransitions(
  current: OrderStatus,
  isAdmin: boolean,
  isWarehouseManager: boolean,
): OrderStatus[] {
  if (isAdmin) return transitions[current];
  if (!isWarehouseManager) return [];
  if (current === OrderStatus.Pending) return [OrderStatus.Processing];
  if (current === OrderStatus.Processing) return [OrderStatus.Shipped];
  return [];
}
